File: build_tenants/python/code/odd_domain/adapters/fpml_confirmation.py
# Implements: REQ-ODD-DOMAIN-BUILD-CAP-001
# Implements: REQ-ODD-DOMAIN-BUILD-CAP-006
# Implements: REQ-ODD-DOMAIN-WORLD-OBJECT-004
# Implements: REQ-ODD-DOMAIN-WORLD-OBJECT-010
"""Bounded FpML confirmation-view parsing for the first trade lane."""

from __future__ import annotations

import xml.etree.ElementTree as ET
from pathlib import Path
from typing import Any


def _local_name(tag: str) -> str:
    if tag.startswith("{"):
        return tag.split("}", 1)[1]
    return tag


def _namespace(root: ET.Element) -> dict[str, str]:
    if root.tag.startswith("{"):
        return {"fpml": root.tag[1:].split("}", 1)[0]}
    return {"fpml": ""}


def _find(element: ET.Element, path: str, ns: dict[str, str]) -> ET.Element | None:
    if ns["fpml"]:
        return element.find(path, ns)
    cleaned = path.replace("fpml:", "")
    return element.find(cleaned)


def _findall(element: ET.Element, path: str, ns: dict[str, str]) -> list[ET.Element]:
    if ns["fpml"]:
        return list(element.findall(path, ns))
    cleaned = path.replace("fpml:", "")
    return list(element.findall(cleaned))


def _text(element: ET.Element | None) -> str | None:
    if element is None or element.text is None:
        return None
    return element.text.strip()


def _text_at(element: ET.Element | None, ns: dict[str, str], *paths: str) -> str | None:
    if element is None:
        return None
    for path in paths:
        value = _text(_find(element, path, ns))
        if value:
            return value
    return None


def _extract_leg(element: ET.Element, ns: dict[str, str], *, index: int) -> dict[str, Any]:
    leg_kind = _local_name(element.tag)
    leg_id = element.attrib.get("id") or f"{leg_kind.lower()}_{index:02d}"
    calculation_schedule_ref = None
    direct_schedule_ref = _find(element, "fpml:calculationPeriodsScheduleReference", ns)
    if direct_schedule_ref is not None:
        calculation_schedule_ref = direct_schedule_ref.attrib.get("href")
    pricing_schedule_ref = _find(
        element,
        "fpml:calculation/fpml:pricingDates/fpml:calculationPeriodsScheduleReference",
        ns,
    )
    if calculation_schedule_ref is None and pricing_schedule_ref is not None:
        calculation_schedule_ref = pricing_schedule_ref.attrib.get("href")

    payer_ref = _find(element, "fpml:payerPartyReference", ns)
    receiver_ref = _find(element, "fpml:receiverPartyReference", ns)
    return {
        "leg_id": leg_id,
        "leg_kind": leg_kind,
        "payer_party_ref": payer_ref.attrib.get("href") if payer_ref is not None else "",
        "receiver_party_ref": receiver_ref.attrib.get("href") if receiver_ref is not None else "",
        "quantity": _text_at(element, ns, "fpml:notionalQuantity/fpml:quantity"),
        "quantity_unit": _text_at(element, ns, "fpml:notionalQuantity/fpml:quantityUnit"),
        "quantity_frequency": _text_at(
            element,
            ns,
            "fpml:notionalQuantity/fpml:quantityFrequency",
        ),
        "total_notional_quantity": _text_at(element, ns, "fpml:totalNotionalQuantity"),
        "fixed_price": _text_at(element, ns, "fpml:fixedPrice/fpml:price"),
        "price_currency": _text_at(element, ns, "fpml:fixedPrice/fpml:priceCurrency"),
        "price_unit": _text_at(element, ns, "fpml:fixedPrice/fpml:priceUnit"),
        "commodity_instrument_id": _text_at(
            element,
            ns,
            "fpml:commodity/fpml:instrumentId",
        ),
        "calculation_schedule_ref": calculation_schedule_ref or "",
    }


def parse_fpml_trade(path: Path) -> dict[str, Any]:
    tree = ET.parse(path)
    root = tree.getroot()
    ns = _namespace(root)

    trade = _find(root, "fpml:trade", ns)
    if trade is None and root.tag.endswith("trade"):
        trade = root
    if trade is None:
        raise ValueError(f"{path}: missing trade element")

    trade_header = _find(trade, "fpml:tradeHeader", ns)
    if trade_header is None:
        raise ValueError(f"{path}: missing tradeHeader")

    trade_ids = [
        trade_id
        for identifier in _findall(trade_header, "fpml:partyTradeIdentifier", ns)
        if (trade_id := _text(_find(identifier, "fpml:tradeId", ns)))
    ]
    trade_id = trade_ids[0] if trade_ids else None
    trade_date = _text(_find(trade_header, "fpml:tradeDate", ns))

    commodity_swap = _find(trade, "fpml:commoditySwap", ns)
    if commodity_swap is None:
        raise ValueError(f"{path}: missing commoditySwap")

    product_type = _text(_find(commodity_swap, "fpml:productType", ns))
    product_id = _text(_find(commodity_swap, "fpml:productId", ns))
    asset_class = _text_at(
        commodity_swap,
        ns,
        "fpml:primaryAssetClass",
        "fpml:assetClass",
    )
    effective_date = _text_at(
        commodity_swap,
        ns,
        "fpml:effectiveDate/fpml:adjustableDate/fpml:unadjustedDate",
        "fpml:effectiveDate",
    )
    termination_date = _text_at(
        commodity_swap,
        ns,
        "fpml:terminationDate/fpml:adjustableDate/fpml:unadjustedDate",
        "fpml:terminationDate",
    )

    party_refs: list[str] = []
    for element in _findall(commodity_swap, ".//fpml:payerPartyReference", ns):
        href = element.attrib.get("href")
        if href and href not in party_refs:
            party_refs.append(href)
    for element in _findall(commodity_swap, ".//fpml:receiverPartyReference", ns):
        href = element.attrib.get("href")
        if href and href not in party_refs:
            party_refs.append(href)

    parties: list[dict[str, str]] = []
    for party in _findall(root, "fpml:party", ns):
        party_xml_id = party.attrib.get("id", "")
        party_id = _text(_find(party, "fpml:partyId", ns)) or party_xml_id
        party_name = _text(_find(party, "fpml:partyName", ns)) or party_id
        parties.append(
            {
                "xml_id": party_xml_id,
                "party_id": party_id,
                "party_name": party_name,
            }
        )

    legs = [
        _extract_leg(element, ns, index=index)
        for index, element in enumerate(list(commodity_swap), start=1)
        if _local_name(element.tag).endswith("Leg")
    ]

    documentation = _find(trade, "fpml:documentation", ns)
    master_agreement = _find(documentation, "fpml:masterAgreement", ns) if documentation is not None else None

    agreement = {
        "type": _text(_find(master_agreement, "fpml:masterAgreementType", ns)),
        "date": _text(_find(master_agreement, "fpml:masterAgreementDate", ns)),
        "version": _text(_find(master_agreement, "fpml:masterAgreementVersion", ns)),
    }

    return {
        "source_standard": "FpML",
        "source_view": "confirmation",
        "fpml_version": root.attrib.get("fpmlVersion", ""),
        "trade_xml_id": trade.attrib.get("id", ""),
        "trade_id": trade_id or "",
        "trade_identifiers": trade_ids,
        "trade_date": trade_date or "",
        "product": {
            "product_xml_id": commodity_swap.attrib.get("id", ""),
            "product_type": product_type or "",
            "product_id": product_id or "",
            "asset_class": asset_class or "",
            "effective_date": effective_date or "",
            "termination_date": termination_date or "",
            "legs": legs,
            "leg_count": len(legs),
        },
        "party_refs": party_refs,
        "parties": parties,
        "agreement": agreement,
    }

File: build_tenants/typescript/code/src/adapters/fpml_confirmation.ts
import { readFile } from "node:fs/promises";
import { XMLParser } from "fast-xml-parser";
import { SyntaxValidator } from "fast-xml-validator";
import type { DeterministicPayloadCheck } from "../domain/semantic_publication.ts";
import type { TypedGap } from "../domain/semantic_memory.ts";
import type { JsonObject } from "../domain/types.ts";

export interface FpmlLegObservation {
  leg_id: string;
  leg_kind: string;
  payer_party_ref: string;
  receiver_party_ref: string;
  quantity: string | null;
  quantity_unit: string | null;
  quantity_frequency: string | null;
  total_notional_quantity: string | null;
  fixed_price: string | null;
  price_currency: string | null;
  price_unit: string | null;
  commodity_instrument_id: string | null;
  calculation_schedule_ref: string;
}

export interface FpmlTradeObservation {
  source_standard: "FpML";
  source_view: "confirmation";
  fpml_version: string;
  trade_xml_id: string;
  trade_id: string;
  trade_identifiers: string[];
  trade_date: string;
  product: {
    product_xml_id: string;
    product_type: string;
    product_id: string;
    asset_class: string;
    effective_date: string;
    termination_date: string;
    legs: FpmlLegObservation[];
    leg_count: number;
  };
  party_refs: string[];
  parties: Array<{ xml_id: string; party_id: string; party_name: string }>;
  agreement: {
    type: string | null;
    date: string | null;
    version: string | null;
  };
}

export const fpmlTradeObservationPayloadCheck: DeterministicPayloadCheck = Object.freeze({
  contractRef: "workspace://build_tenants/common/schemas/fpml_trade_observation.schema.json",
  validate: (payload: JsonObject): readonly TypedGap[] => {
    const candidate = payload as unknown as Partial<FpmlTradeObservation>;
    if (candidate.source_standard !== "FpML" || candidate.source_view !== "confirmation") {
      throw new Error("FpML observation must declare the FpML confirmation source view");
    }
    if (typeof candidate.fpml_version !== "string" || candidate.fpml_version.length === 0) {
      throw new Error("FpML observation has no version");
    }
    if (typeof candidate.trade_id !== "string" || candidate.trade_id.length === 0) {
      throw new Error("FpML observation has no trade identity");
    }
    if (typeof candidate.trade_date !== "string" || candidate.trade_date.length === 0) {
      throw new Error("FpML observation has no trade date");
    }
    const product = candidate.product;
    if (product === undefined || !Array.isArray(product.legs) || product.legs.length === 0) {
      throw new Error("FpML observation has no product legs");
    }
    if (product.leg_count !== product.legs.length) {
      throw new Error("FpML observation leg_count does not match its legs");
    }
    return Object.freeze([]);
  }
});

type XmlRecord = Record<string, unknown>;

const repeatedFpmlElements = new Set([
  "partyTradeIdentifier",
  "party",
  "fixedLeg",
  "floatingLeg"
]);

const fpmlParser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: "@_",
  removeNSPrefix: true,
  parseTagValue: false,
  parseAttributeValue: false,
  trimValues: true,
  processEntities: true,
  isArray: (tagName) => repeatedFpmlElements.has(tagName)
});

function xmlRecord(value: unknown, label: string): XmlRecord {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${label} must be an XML element`);
  }
  return value as XmlRecord;
}

function optionalXmlRecord(value: unknown): XmlRecord | null {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as XmlRecord
    : null;
}

function xmlRecords(value: unknown, label: string): readonly XmlRecord[] {
  if (value === undefined) return [];
  const values = Array.isArray(value) ? value : [value];
  return values.map((item, index) => xmlRecord(item, `${label}[${index}]`));
}

function xmlText(value: unknown): string | null {
  if (typeof value === "string") return value.length > 0 ? value : null;
  const record = optionalXmlRecord(value);
  const text = record?.["#text"];
  return typeof text === "string" && text.length > 0 ? text : null;
}

function xmlAttribute(value: unknown, name: string): string {
  const attribute = optionalXmlRecord(value)?.[`@_${name}`];
  return typeof attribute === "string" ? attribute : "";
}

function xmlHref(value: unknown): string {
  return xmlAttribute(value, "href");
}

function nestedRecord(parent: XmlRecord | null, name: string): XmlRecord | null {
  return optionalXmlRecord(parent?.[name]);
}

function extractLeg(leg: XmlRecord, legKind: string, index: number): FpmlLegObservation {
  const notional = nestedRecord(leg, "notionalQuantity");
  const fixedPrice = nestedRecord(leg, "fixedPrice");
  const commodity = nestedRecord(leg, "commodity");
  const calculation = nestedRecord(leg, "calculation");
  const pricingDates = nestedRecord(calculation, "pricingDates");
  return {
    leg_id: xmlAttribute(leg, "id") || `${legKind.toLowerCase()}_${String(index).padStart(2, "0")}`,
    leg_kind: legKind,
    payer_party_ref: xmlHref(leg.payerPartyReference),
    receiver_party_ref: xmlHref(leg.receiverPartyReference),
    quantity: xmlText(notional?.quantity),
    quantity_unit: xmlText(notional?.quantityUnit),
    quantity_frequency: xmlText(notional?.quantityFrequency),
    total_notional_quantity: xmlText(leg.totalNotionalQuantity),
    fixed_price: xmlText(fixedPrice?.price),
    price_currency: xmlText(fixedPrice?.priceCurrency),
    price_unit: xmlText(fixedPrice?.priceUnit),
    commodity_instrument_id: xmlText(commodity?.instrumentId),
    calculation_schedule_ref: xmlHref(leg.calculationPeriodsScheduleReference) ||
      xmlHref(pricingDates?.calculationPeriodsScheduleReference)
  };
}

function unique(values: string[]): string[] {
  const seen = new Set<string>();
  const result: string[] = [];
  for (const value of values) {
    if (value && !seen.has(value)) {
      seen.add(value);
      result.push(value);
    }
  }
  return result;
}

export async function parseFpmlTrade(filePath: string): Promise<FpmlTradeObservation> {
  const xml = await readFile(filePath, "utf8");
  let syntax: ReturnType<typeof SyntaxValidator.validate>;
  try {
    syntax = SyntaxValidator.validate(xml, {
      docType: { maxEntityCount: 0, maxEntitySize: 0 }
    });
  } catch (error: unknown) {
    throw new Error(`${filePath}: malformed XML: ${error instanceof Error ? error.message : String(error)}`);
  }
  if (syntax !== true) {
    throw new Error(
      `${filePath}: malformed XML at ${syntax.err.line}:${syntax.err.col}: ${syntax.err.msg}`
    );
  }
  const parsed = fpmlParser.parse(xml) as unknown;
  const document = xmlRecord(xmlRecord(parsed, filePath).dataDocument, `${filePath}: dataDocument`);
  const trade = xmlRecord(document.trade, `${filePath}: trade`);
  const tradeHeader = xmlRecord(trade.tradeHeader, `${filePath}: tradeHeader`);
  const tradeIdentifiers = xmlRecords(
    tradeHeader.partyTradeIdentifier,
    `${filePath}: partyTradeIdentifier`
  )
    .map((identifier) => xmlText(identifier.tradeId))
    .filter((value): value is string => typeof value === "string" && value.length > 0);
  const commoditySwap = xmlRecord(trade.commoditySwap, `${filePath}: commoditySwap`);
  const fixedLegs = xmlRecords(commoditySwap.fixedLeg, `${filePath}: fixedLeg`);
  const floatingLegs = xmlRecords(commoditySwap.floatingLeg, `${filePath}: floatingLeg`);
  const legs = [
    ...fixedLegs.map((leg, index) => extractLeg(leg, "fixedLeg", index + 1)),
    ...floatingLegs.map((leg, index) => extractLeg(leg, "floatingLeg", index + 1 + fixedLegs.length))
  ];
  const partyRefs = unique(legs.flatMap((leg) => [leg.payer_party_ref, leg.receiver_party_ref]));
  const parties = xmlRecords(document.party, `${filePath}: party`).map((party) => {
    const xmlId = xmlAttribute(party, "id");
    const partyId = xmlText(party.partyId) || xmlId;
    return {
      xml_id: xmlId,
      party_id: partyId,
      party_name: xmlText(party.partyName) || partyId
    };
  });
  const masterAgreement = nestedRecord(nestedRecord(trade, "documentation"), "masterAgreement");
  const effectiveDate = nestedRecord(commoditySwap, "effectiveDate");
  const terminationDate = nestedRecord(commoditySwap, "terminationDate");
  return {
    source_standard: "FpML",
    source_view: "confirmation",
    fpml_version: xmlAttribute(document, "fpmlVersion"),
    trade_xml_id: xmlAttribute(trade, "id"),
    trade_id: tradeIdentifiers[0] ?? "",
    trade_identifiers: tradeIdentifiers,
    trade_date: xmlText(tradeHeader.tradeDate) ?? "",
    product: {
      product_xml_id: xmlAttribute(commoditySwap, "id"),
      product_type: xmlText(commoditySwap.productType) ?? "",
      product_id: xmlText(commoditySwap.productId) ?? "",
      asset_class: xmlText(commoditySwap.primaryAssetClass) ?? xmlText(commoditySwap.assetClass) ?? "",
      effective_date: xmlText(nestedRecord(effectiveDate, "adjustableDate")?.unadjustedDate) ??
        xmlText(effectiveDate?.unadjustedDate) ?? xmlText(commoditySwap.effectiveDate) ?? "",
      termination_date: xmlText(nestedRecord(terminationDate, "adjustableDate")?.unadjustedDate) ??
        xmlText(terminationDate?.unadjustedDate) ?? xmlText(commoditySwap.terminationDate) ?? "",
      legs,
      leg_count: legs.length
    },
    party_refs: partyRefs,
    parties,
    agreement: {
      type: xmlText(masterAgreement?.masterAgreementType),
      date: xmlText(masterAgreement?.masterAgreementDate),
      version: xmlText(masterAgreement?.masterAgreementVersion)
    }
  };
}
